imbi/app.py

```diff
@@ -181,3 +190,107 @@ def validate_request(self, request: httputil.HTTPServerRequest) -> None:
+        the feature is enabled *and* the fact type ID is configured.
+
+        """
+        config = self.settings.get('component_scoring', {})
+        if not config.get('enabled'):
+            config['enabled'] = False  # ensure that `enabled' key exists
+            config['project_fact_type_id'] = None  # make this consistent too
+            return
+
+        component_fact_id: int | None = config.get('project_fact_type_id')
+        if component_fact_id is not None:
+            return
+
+        fact_name = config.get('fact_name', 'Component Score')
+        result = await transaction.execute(
+            'SELECT id'
+            '  FROM v1.project_fact_types'
+            ' WHERE name = %(fact_name)s', {'fact_name': fact_name},
+            metric_name='get-project-fact-type')
+        if result:
+            self.logger.debug('found %r project fact type with id %r',
+                              fact_name, result.row['id'])
+            config['project_fact_type_id'] = result.row['id']
+            return
+
+        self.logger.debug('creating Component Score fact type named %r',
+                          fact_name)
+        result = await transaction.execute(
+            'INSERT INTO v1.project_fact_types('
+            '              created_by, name, fact_type, data_type,'
+            '              project_type_ids, ui_options)'
+            '     VALUES (%(created_by)s, %(name)s, %(fact_type)s,'
+            '             %(data_type)s, %(empty_array)s, %(ui_options)s)'
+            '  RETURNING id', {
+                'empty_array': [],
+                'ui_options': ['read-only'],
+                'created_by': 'system',
+                'fact_type': 'range',
+                'data_type': 'integer',
+                'name': fact_name,
+            })
+
+        # Create range values that encompass the ProjectStatus enum
+        # values in imbi.endpoints.components.models. We want a wide
+        # range in the center for flexibility in introducing more
+        # "yellow/orange" categories in the future.
+        config['project_fact_type_id'] = result.row['id']
+        last_value = 0.0
+        for max_value in (
+                20,  # arbitrary low-end of middle range
+                imbi.endpoints.components.models.ProjectStatus.NEEDS_WORK,
+                imbi.endpoints.components.models.ProjectStatus.OKAY,
+        ):
+            await transaction.execute(
+                'INSERT INTO v1.project_fact_type_ranges('
+                '               created_by, fact_type_id, min_value,'
+                '               max_value, score)'
+                '     VALUES (%(created_by)s, %(fact_id)s, %(min_value)s,'
+                '             %(max_value)s, %(max_value)s'
+                ')', {
+                    'created_by': 'system',
+                    'fact_id': config['project_fact_type_id'],
+                    'min_value': last_value,
+                    'max_value': max_value,
```

Comment:
I wonder if we actually want unacceptable to be score of 0 instead of 20?

Reply:
I poked around locally and I think that I can justify having a non-zero score. The way that the scoring algorithm works, a scored fact has a value of zero if it is absent for the project. Having a non-zero score here for the worst value makes it slightly better to have a score than not have one?! maybe.....

With a local project with a somewhat arbitrary collection of data... the fact being absent gave be a score of 62.5. When the fact is present with a value of 20, the score is 66.25. I have the weight set to 30/160 FWIW.

Reply:
The only reason that I had it this way was ease of implementation. All of the range values use the upper-value of the range as the score.